### PR TITLE
Catalog promotions action

### PR DESCRIPTION
**Pull request by SirDomin**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | 
| Deprecations?   | no
| Related tickets | 
| License         | MIT

### CODE REVIEW

File: src/Sylius/Bundle/CoreBundle/Migrations/Version20210826063828.php [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+use Doctrine\Migrations\AbstractMigration;
+
+/**
+ * Auto-generated Migration: Please modify to your needs!
+ */
+final class Version20210826063828 extends AbstractMigration
+{
+    public function getDescription(): string
+    {
+        return '';
```
Comment (AdamKasp) [inline src/Sylius/Bundle/CoreBundle/Migrations/Version20210826063828.php:17]:
```suggestion
        return 'Add catalog promotion actions';
```

File: src/Sylius/Component/Promotion/Model/CatalogPromotionInterface.php [hunk 2]
```diff
@@ -21,6 +21,10 @@
 
 interface CatalogPromotionInterface extends ResourceInterface, CodeAwareInterface, TranslatableInterface
 {
+    public const TYPE_PERCENTAGE_DISCOUNT = 'percentage_product_discount';
```
Comment (AdamKasp) [inline src/Sylius/Component/Promotion/Model/CatalogPromotionInterface.php:24]: Why did you change the place of these costs? shouldn't we have types of rules on the rule Interface and types of actions on the Action interface?

File: features/promotion/managing_catalog_promotions/editing_catalog_promotion.feature [hunk 3]
```diff
@@ -49,3 +49,9 @@ Feature: Editing catalog promotion
         When I want "Christmas sale" catalog promotion to be applied on "Kotlin T-shirt" variant
         Then I should be notified that it has been successfully edited
         And this catalog promotion should be applied on "Kotlin T-shirt" variant
+
+    @api
+    Scenario: Editing catalog promotion action
+        When I want "Christmas sale" catalog promotion to have "40%" discount
```
Comment (GSadee) [inline features/promotion/managing_catalog_promotions/editing_catalog_promotion.feature:55]:
```suggestion
        When I edit "Christmas sale" catalog promotion to have "40%" discount
```

File: src/Sylius/Behat/Context/Api/Admin/ManagingCatalogPromotionsContext.php [hunk 4]
```diff
@@ -169,6 +171,23 @@ public function iWantToModifyACatalogPromotion(CatalogPromotionInterface $catalo
...
+    public function iAddThePercentageDiscountCatalogPromotionActionConfiguredWithAmountOf(string $amount): void
+    {
+        $actions = [
+            [
+                'type' => CatalogPromotionInterface::TYPE_PERCENTAGE_DISCOUNT,
+                'configuration' => [
+                    'amount' => $amount
+                ],
+            ],
+        ];
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Admin/ManagingCatalogPromotionsContext.php:186]:
```suggestion
        $actions = [[
            'type' => CatalogPromotionInterface::TYPE_PERCENTAGE_DISCOUNT,
            'configuration' => [
                'amount' => $amount
            ],
        ]];
```

File: src/Sylius/Behat/Context/Api/Admin/ManagingCatalogPromotionsContext.php [hunk 5]
```diff
@@ -230,6 +266,35 @@ public function thereShouldBeNewCatalogPromotionOnTheList(int $amount = 0): void
         Assert::count($this->responseChecker->getCollection($this->client->index()), $amount);
     }
 
+    /**
+     * @Then I should be notified that it has been successfully created
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Admin/ManagingCatalogPromotionsContext.php:270]: Do you need this step? I don't see you adding it

File: src/Sylius/Bundle/ApiBundle/Resources/config/serialization/CatalogPromotion.xml [hunk 6]
```diff
@@ -43,11 +43,13 @@
...
             <group>admin:catalog_promotion:read</group>
             <group>admin:catalog_promotion:create</group>
+            <group>admin:catalog_promotion:update</group>
         </attribute>
-        <attribute name="rules">
+        <attribute name="actions">
             <group>admin:catalog_promotion:read</group>
             <group>admin:catalog_promotion:create</group>
             <group>admin:catalog_promotion:update</group>
+
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/serialization/CatalogPromotion.xml:52]:
```suggestion
```

File: src/Sylius/Bundle/PromotionBundle/Resources/config/doctrine/model/CatalogPromotionAction.orm.xml [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+-->
+
+<doctrine-mapping xmlns="http://doctrine-project.org/schemas/orm/doctrine-mapping" xmlns:gedmo="http://gediminasm.org/schemas/orm/doctrine-extensions-mapping">
+    <mapped-superclass name="Sylius\Component\Promotion\Model\CatalogPromotionAction" table="sylius_catalog_promotion_action">
+        <id name="id" type="integer">
+            <generator strategy="AUTO" />
+        </id>
+
+        <field name="type" type="string" />
```
Comment (GSadee) [inline src/Sylius/Bundle/PromotionBundle/Resources/config/doctrine/model/CatalogPromotionAction.orm.xml:20]:
```suggestion
        <field name="type" />
```

File: src/Sylius/Component/Promotion/Model/CatalogPromotion.php [hunk 8]
```diff
@@ -39,11 +39,15 @@ class CatalogPromotion implements CatalogPromotionInterface
...
     protected Collection $rules;
 
+    protected Collection $actions;
+
     public function __construct()
     {
         $this->initializeTranslationsCollection();
 
         $this->rules = new ArrayCollection();
+
```
Comment (GSadee) [inline src/Sylius/Component/Promotion/Model/CatalogPromotion.php:49]:
```suggestion
```

File: src/Sylius/Component/Promotion/Model/CatalogPromotionInterface.php (hunk 2 above)
Comment (SirDomin) [reply src/Sylius/Component/Promotion/Model/CatalogPromotionInterface.php]: it wasnt needed :D

File: src/Sylius/Bundle/PromotionBundle/DependencyInjection/Configuration.php [hunk 9]
```diff
@@ -148,6 +150,23 @@ private function addResourcesSection(ArrayNodeDefinition $node): void
...
                         ->end()
+                        ->arrayNode('catalog_promotion_action')
+                            ->addDefaultsIfNotSet()
+                            ->children()
+                                ->variableNode('options')->end()
+                                ->arrayNode('classes')
+                                    ->addDefaultsIfNotSet()
+                                    ->children()
+                                        ->scalarNode('model')->defaultValue(CatalogPromotionAction::class)->cannotBeEmpty()->end()
+                                        ->scalarNode('interface')->defaultValue(CatalogPromotionActionInterface::class)->cannotBeEmpty()->end()
```
Comment (loevgaard) [inline src/Sylius/Bundle/PromotionBundle/DependencyInjection/Configuration.php:161]: Isn't needed anymore

```suggestion
```

File: src/Sylius/Behat/Context/Api/Admin/ManagingCatalogPromotionsContext.php [hunk 10]
```diff
@@ -193,7 +210,7 @@ public function iBrowseCatalogPromotions(): void
     public function iAddTheRuleConfiguredWithProductAnd(ProductVariantInterface $firstVariant, ProductVariantInterface $secondVariant): void
     {
         $rules = [[
-            'type' => CatalogPromotionRuleInterface::TYPE_CONTAINS_VARIANTS,
+            'type' => CatalogPromotionInterface::TYPE_CONTAINS_VARIANTS,
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Admin/ManagingCatalogPromotionsContext.php:213]: This type should be on `CatalogPromotionRuleInterface`

File: src/Sylius/Behat/Context/Api/Admin/ManagingCatalogPromotionsContext.php [hunk 11]
```diff
@@ -230,6 +264,27 @@ public function thereShouldBeNewCatalogPromotionOnTheList(int $amount = 0): void
...
     }
 
+    /**
+     * @Then /^it should have "([^"]+)" discount$/
+     */
+    public function itShouldHaveDiscount(string $amount): void
+    {
+        $catalogPromotion = $this->responseChecker->getCollection($this->client->getLastResponse())[0];
+
+        Assert::same($catalogPromotion['actions'][0]['configuration']['amount'], $amount);
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Api/Admin/ManagingCatalogPromotionsContext.php:274]: Now the question is - should we use typical percentage notation (40% -> 0.4) or absolute values(40% -> 40)? For sure, we shouldn't use `40%` as a config value. This transformation should be done before assertion

File: src/Sylius/Behat/Context/Api/Admin/ManagingCatalogPromotionsContext.php [hunk 12]
```diff
@@ -230,6 +264,27 @@ public function thereShouldBeNewCatalogPromotionOnTheList(int $amount = 0): void
         Assert::count($this->responseChecker->getCollection($this->client->index()), $amount);
     }
 
+    /**
+     * @Then /^it should have "([^"]+)" discount$/
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Api/Admin/ManagingCatalogPromotionsContext.php:268]: this step does not suggest that it will be performed only on the first item from the collection. I would expect the same behaviour as in the step below for this one

File: tests/Api/Admin/CatalogPromotionsTest.php [hunk 13]
```diff
@@ -166,9 +174,17 @@ public function it_updates_catalog_promotion(): void
             json_encode([
                 'name' => 'T-Shirts discount',
                 'code' => 'new_code',
+                'actions' => [
+                    [
+                        'type' => CatalogPromotionActionInterface::TYPE_PERCENTAGE_DISCOUNT,
+                        'configuration' => [
+                            'amount' => '40%'
```
Comment (lchrusciel) [inline tests/Api/Admin/CatalogPromotionsTest.php:181]:
```suggestion
                            'amount' => 40
```
or
```suggestion
                            'amount' => 0.4
```

I prefer the latter

File: tests/Api/Responses/Expected/admin/catalog_promotion/post_catalog_promotion_response.json [hunk 14]
```diff
@@ -19,6 +19,17 @@
             ]
         }
     ],
+    "actions": [
+        {
+            "@type": "CatalogPromotionAction",
+            "@id": @string@,
```
Comment (lchrusciel) [inline tests/Api/Responses/Expected/admin/catalog_promotion/post_catalog_promotion_response.json:25]:
```suggestion
            "@id": "\/api\/v2\/admin\/catalog-promotion-actions\/@integer@",
```
Won't it work?

Comment (GSadee) [reply tests/Api/Responses/Expected/admin/catalog_promotion/post_catalog_promotion_response.json:25]: I had a similar observation in PR with rules, but if I'm correct, there is no get endpoint for CatalogPromotionAction, so it doesn't have IRI. We are operating on rules and actions only by operating on CatalogPromotion entity

**Review by Arminek — APPROVED** [review]

File: src/Sylius/Behat/Context/Api/Admin/ManagingCatalogPromotionsContext.php (hunk 11 above)
Comment (Zales0123) [reply src/Sylius/Behat/Context/Api/Admin/ManagingCatalogPromotionsContext.php:274]: I would go with `0.4` as it's consistent with actions on cart promotions and is easier to be used for calculation

File: src/Sylius/Behat/Context/Api/Admin/ManagingCatalogPromotionsContext.php (hunk 12 above)
Comment (Zales0123) [reply src/Sylius/Behat/Context/Api/Admin/ManagingCatalogPromotionsContext.php:268]: Indeed - I assume it's inspired by `it should apply to "PHP T-Shirt" variant and "Kotlin T-Shirt" variant` step. However, it makes sense in the scenario 😄 So maybe I would leave it for now a try to refactor it in the separate PR?

File: src/Sylius/Bundle/PromotionBundle/DependencyInjection/Configuration.php (hunk 9 above)
Comment (Zales0123) [reply src/Sylius/Bundle/PromotionBundle/DependencyInjection/Configuration.php:163]: Indeed 👍 but to be consistent with other configurations I would leave it for now and remove altogether later 🖖

**Comment by GSadee** [comment]
Thanks, @SirDomin! :1st_place_medal:
